File: examples/windows-maui/Tests/OrderPageTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using UITestProbe.Net;

namespace OrderApp.Tests;

[TestClass]
public class OrderPageTests
{
    private ProbeDriver probe = null!;

    [TestInitialize]
    public void Setup()
    {
        var app = AppDriver.Launch<App>();
        probe = new ProbeDriver(app);
    }

    // -- Page Load --

    [TestMethod]
    public void PageReady()
    {
        probe.WaitForPageReady(timeoutMs: 5000);

        var page = probe.Query("order-page");
        Assert.AreEqual(ProbeType.Page, page.Type);
        Assert.IsTrue(page.IsReady);
    }

    [TestMethod]
    public void OrderTableLoaded()
    {
        probe.WaitForPageReady();

        var table = probe.Query("order-table");
        Assert.AreEqual(ProbeType.DataContainer, table.Type);
        Assert.AreEqual(ProbeState.Loaded, table.State);
        Assert.IsTrue(table.ChildCount > 0, "Table should have rows");
        Assert.AreEqual("GET /api/orders", table.Source);
    }

    // -- Filter Linkage --

    [TestMethod]
    public void StatusFilterReloadsTable()
    {
        probe.WaitForPageReady();

        probe.ActAndWait(
            probeId: "status-filter",
            action: "select:completed",
            target: "order-table",
            expectedState: "loaded"
        );

        var table = probe.Query("order-table");
        Assert.AreEqual(ProbeState.Loaded, table.State);
    }

    [TestMethod]
    public void FilterLinkageMetadata()
    {
        probe.WaitForPageReady();

        var linkage = probe.VerifyLinkage(
            probeId: "status-filter",
            action: "select:completed"
        );
        Assert.AreEqual("order-table", linkage.Target);
        Assert.AreEqual(LinkageEffect.DataReload, linkage.Effect);
        Assert.AreEqual("GET /api/orders", linkage.ApiPath);
    }

    // -- Pagination --

    [TestMethod]
    public void PaginatorNavigation()
    {
        probe.WaitForPageReady();

        probe.ActAndWait(
            probeId: "paginator",
            action: "tap:next",
            target: "order-table",
            expectedState: "loaded"
        );

        var table = probe.Query("order-table");
        Assert.AreEqual(ProbeState.Loaded, table.State);
        Assert.IsTrue(table.ChildCount > 0);
    }

    // -- Create Order Modal --

    [TestMethod]
    public void CreateOrderFlow()
    {
        probe.WaitForPageReady();

        probe.Tap("create-order-btn");
        probe.WaitFor("create-order-modal", ProbeState.Visible);

        var modal = probe.Query("create-order-modal");
        Assert.AreEqual(ProbeType.Modal, modal.Type);

        probe.Fill("input-customer", "Acme Corp");
        probe.Fill("input-amount", "1500.00");

        probe.ActAndWait(
            probeId: "submit-order-btn",
            action: "tap",
            target: "order-table",
            expectedState: "loaded"
        );

        Assert.IsFalse(probe.IsVisible("create-order-modal"));
    }

    [TestMethod]
    public void FormValidation()
    {
        probe.WaitForPageReady();
        probe.Tap("create-order-btn");
        probe.WaitFor("create-order-modal", ProbeState.Visible);

        probe.Tap("submit-order-btn");

        var customerField = probe.Query("input-customer");
        Assert.IsTrue(customerField.HasValidationError);
        Assert.AreEqual("Customer is required", customerField.ValidationMessage);
    }

    // -- Visibility --

    [TestMethod]
    public void ElementVisibility()
    {
        probe.WaitForPageReady();

        Assert.IsTrue(probe.IsEffectivelyVisible("order-table"));
        Assert.IsTrue(probe.IsEffectivelyVisible("status-filter"));
        Assert.IsFalse(probe.IsEffectivelyVisible("create-order-modal"));
    }

    // -- Device Presets --

    [TestMethod]
    public void ResponsiveLayout_Desktop()
    {
        probe.SetDevice(DevicePreset.Desktop1080p);
        probe.WaitForPageReady();

        Assert.IsTrue(probe.IsEffectivelyVisible("order-table"));
        Assert.IsTrue(probe.IsEffectivelyVisible("status-filter"));
    }

    [TestMethod]
    public void ResponsiveLayout_DesktopWide()
    {
        probe.SetDevice(DevicePreset.Desktop1440p);
        probe.WaitForPageReady();

        Assert.IsTrue(probe.IsEffectivelyVisible("order-table"));
        Assert.IsTrue(probe.IsEffectivelyVisible("paginator"));
    }
}

File: examples/ios-swiftui/Tests/OrderPageTests.swift
import XCTest
import UITestProbe

final class OrderPageTests: XCTestCase {
    var app: XCUIApplication!
    var probe: ProbeXCTest!

    override func setUpWithError() throws {
        continueAfterFailure = false
        app = XCUIApplication()
        app.launch()
        probe = ProbeXCTest(app)
    }

    // MARK: - Page Load

    func testPageReady() throws {
        probe.waitForPageReady(timeout: 5)

        let page = probe.query("order-page")
        XCTAssertEqual(page.type, .page)
        XCTAssertTrue(page.isReady)
    }

    func testOrderTableLoaded() throws {
        probe.waitForPageReady()

        let table = probe.query("order-table")
        XCTAssertEqual(table.type, .dataContainer)
        XCTAssertEqual(table.state, .loaded)
        XCTAssertGreaterThan(table.childCount, 0, "Table should have rows")
        XCTAssertEqual(table.source, "GET /api/orders")
    }

    // MARK: - Filter Linkage

    func testStatusFilterReloadsTable() throws {
        probe.waitForPageReady()

        // Act: select a filter value, wait for the linked target to reload
        probe.actAndWait(
            "status-filter",
            action: .select("completed"),
            target: "order-table",
            state: "loaded"
        )

        let table = probe.query("order-table")
        XCTAssertEqual(table.state, .loaded)
    }

    func testFilterLinkageMetadata() throws {
        probe.waitForPageReady()

        let linkage = probe.verifyLinkage(
            "status-filter",
            action: .select("completed")
        )
        XCTAssertEqual(linkage.target, "order-table")
        XCTAssertEqual(linkage.effect, .dataReload)
        XCTAssertEqual(linkage.apiPath, "GET /api/orders")
    }

    // MARK: - Pagination

    func testPaginatorNavigation() throws {
        probe.waitForPageReady()

        let firstRows = probe.query("order-table").childCount

        probe.actAndWait(
            "paginator",
            action: .tap("next"),
            target: "order-table",
            state: "loaded"
        )

        let table = probe.query("order-table")
        XCTAssertEqual(table.state, .loaded)
        // Page changed so content may differ
        XCTAssertGreaterThan(table.childCount, 0)
    }

    // MARK: - Create Order Modal

    func testCreateOrderFlow() throws {
        probe.waitForPageReady()

        // Open modal
        probe.tap("create-order-btn")
        probe.waitFor("create-order-modal", state: .visible)

        let modal = probe.query("create-order-modal")
        XCTAssertEqual(modal.type, .modal)

        // Fill form
        probe.fill("input-customer", value: "Acme Corp")
        probe.fill("input-amount", value: "1500.00")

        // Submit and wait for table reload
        probe.actAndWait(
            "submit-order-btn",
            action: .tap(),
            target: "order-table",
            state: "loaded"
        )

        // Modal should be dismissed
        XCTAssertFalse(probe.isVisible("create-order-modal"))
    }

    func testFormValidation() throws {
        probe.waitForPageReady()
        probe.tap("create-order-btn")
        probe.waitFor("create-order-modal", state: .visible)

        // Submit with empty fields
        probe.tap("submit-order-btn")

        let customerField = probe.query("input-customer")
        XCTAssertTrue(customerField.hasValidationError)
        XCTAssertEqual(customerField.validationMessage, "Customer is required")
    }

    // MARK: - Visibility

    func testElementVisibility() throws {
        probe.waitForPageReady()

        XCTAssertTrue(probe.isEffectivelyVisible("order-table"))
        XCTAssertTrue(probe.isEffectivelyVisible("status-filter"))
        XCTAssertFalse(probe.isEffectivelyVisible("create-order-modal"))
    }

    // MARK: - Device Presets

    func testResponsiveLayout_iPhone() throws {
        probe.setDevice(.iPhone15Pro)
        probe.waitForPageReady()

        XCTAssertTrue(probe.isEffectivelyVisible("order-table"))
        XCTAssertTrue(probe.isEffectivelyVisible("paginator"))
    }

    func testResponsiveLayout_iPad() throws {
        probe.setDevice(.iPadAir)
        probe.waitForPageReady()

        XCTAssertTrue(probe.isEffectivelyVisible("order-table"))
        XCTAssertTrue(probe.isEffectivelyVisible("status-filter"))
    }
}
